libs/vkd3d/va_map.c

```diff
@@ -0,0 +1,382 @@
+    {
+        const struct vkd3d_unique_resource *r;
+        size_t i = lo + (hi - lo) / 2;
+
+        r = va_map->small_entries[i];
+
+        if (va < r->va)
+            hi = i;
+        else if (va >= r->va + r->size)
+            lo = i + 1;
+        else
+        {
+            lo = hi = i;
+            resource = r;
+        }
+    }
+
+    if (index)
+        *index = lo;
+
+    return resource;
+}
+
+void vkd3d_va_map_insert(struct vkd3d_va_map *va_map, const struct vkd3d_unique_resource *resource)
+{
+    VkDeviceAddress block_va, min_va, max_va;
+    struct vkd3d_va_block *block;
+    size_t index;
+
+    if (resource->size >= VKD3D_VA_BLOCK_SIZE)
+    {
+        min_va = resource->va;
+        max_va = resource->va + resource->size;
+        block_va = min_va & ~VKD3D_VA_LO_MASK;
+
+        while (block_va < max_va)
+        {
+            block = vkd3d_va_map_get_block(va_map, block_va);
+
+            if (block_va < min_va)
+            {
+                block->r.va = min_va;
+                block->r.resource = resource;
+            }
+            else
+            {
+                block->l.va = max_va;
+                block->l.resource = resource;
+            }
+
+            block_va += VKD3D_VA_BLOCK_SIZE;
+        }
+    }
+    else
+    {
+        pthread_mutex_lock(&va_map->mutex);
+
+        if (!vkd3d_va_map_find_entry(va_map, resource->va, &index))
+        {
+            vkd3d_array_reserve((void**)&va_map->small_entries, &va_map->small_entries_size,
+                    va_map->small_entries_count + 1, sizeof(*va_map->small_entries));
+
+            memmove(&va_map->small_entries[index + 1], &va_map->small_entries[index],
+                    sizeof(*va_map->small_entries) * (va_map->small_entries_count - index));
+
+            va_map->small_entries[index] = resource;
+            va_map->small_entries_count += 1;
+        }
+
+        pthread_mutex_unlock(&va_map->mutex);
+    }
+}
+
+void vkd3d_va_map_remove(struct vkd3d_va_map *va_map, const struct vkd3d_unique_resource *resource)
+{
+    VkDeviceAddress block_va, min_va, max_va;
+    struct vkd3d_va_block *block;
+    size_t index;
+
+    if (resource->size >= VKD3D_VA_BLOCK_SIZE)
+    {
+        min_va = resource->va;
+        max_va = resource->va + resource->size;
+        block_va = min_va & ~VKD3D_VA_LO_MASK;
+
+        while (block_va < max_va)
+        {
+            block = vkd3d_va_map_get_block(va_map, block_va);
+
+            if (block->l.resource == resource)
+            {
+                block->l.va = 0;
+                block->l.resource = NULL;
+            }
+            else if (block->r.resource == resource)
+            {
+                block->r.va = 0;
+                block->r.resource = NULL;
+            }
+
+            block_va += VKD3D_VA_BLOCK_SIZE;
+        }
+    }
+    else
+    {
+        pthread_mutex_lock(&va_map->mutex);
+
+        if (vkd3d_va_map_find_entry(va_map, resource->va, &index) == resource)
```

Comment:
Should probably call this find_small_entry to make it more distinct?

Reply:
Fixed.